### PR TITLE
refined labeling

### PR DESCRIPTION
**Pull request by cgcostume**
* [x] dynamic labels are transformed statically and dynamically
* [x] the dynamic transformations is set per label as uniform
* [x] dynamic labels are not drawn with single call but dispatched per label (in order to change uniforms)
* [x] labels themselves must decide where to put a all relevant transforms, e.g., a 2d screenspace label with size in pt will use static baking and re-typesetting etc. when frame size changes - the dynamic label can provide an additional scale that is applied in the vertex shader. No typesetting required.
* [x] decided against instanced with texture lookup (16 float32 texel fetches with no easy fallback for ie11)
* [x] issues to be fixed: macOS safari label rendering (also ie11)
* [x] major memory/perf improvement, since for now, typeset results are cached and computed in-place instead of multiple vertex to vec to vertex to float32 conversions.

### CODE REVIEW

**Comment by cgcostume** [comment]
giphy

**Comment by cgcostume** [comment]
tested on Windows (IE11, Edge, Chrome, Fire Fox) and macOS (Chrome, Safari) - worked.
The dynamic placement is only rudimentary for now and should be enhanced based on actual use.

File: website/js/website.js [hunk 1]
```diff
@@ -13,7 +13,7 @@ window.onload = function () {
     aboutCode = window.document.getElementById('context-about');
     aboutCode.innerText = context.aboutString();
 
-    canvas.controller.multiFrameNumber = 8;
+    canvas.controller.multiFrameNumber = 1;
```
Comment (scheibel) [inline website/js/website.js]: Is this change intended or is this the result of a debug session?

**Comment by scheibel** [comment]
This feature PR is hard to review as there is much orthogonal code work (node update, linter appeasement)

File: website/js/website.js (hunk 1 above)
Comment (cgcostume) [reply website/js/website.js]: debug session and ok for now, since the default example is always adjusted to its current needs

File: source/text/glyph.vert [hunk 2]
```diff
@@ -5,25 +5,30 @@ precision lowp int;
...
-    layout(location = 2) in vec3 a_origin;
-    layout(location = 3) in vec3 a_tangent;
-    layout(location = 4) in vec3 a_up;
+
+in vec2 a_vertex;
+/* [ texture ll: vec2, ur: vec2 ]*/
+in vec4 a_texCoord;
+in vec3 a_origin;
+in vec3 a_tangent;
+in vec3 a_bitangent;
```
Comment (anne-gropler) [inline source/text/glyph.vert]: I am not sure if `bitangent` is the correct word, see [bitanget](https://en.wikipedia.org/wiki/Bitangent). This is why, some time ago, we changed `bitan` to `up`.

File: source/text/glyphvertices.ts [hunk 3]
```diff
@@ -140,10 +157,10 @@ export class GlyphVertices {
     }
 
     /**
-     * All GlyphVertex up vectors gathered in one buffer. Call update() to update it.
+     * All GlyphVertex up/bitangent vectors gathered in one buffer. Call update() to update it.
```
Comment (anne-gropler) [inline source/text/glyphvertices.ts]: doc: there is no `update()` anymore. Same for origins, tangents, bitangents and texCoords.

File: source/text/label.ts [hunk 4]
```diff
@@ -77,8 +84,11 @@ export class Label {
 
 
     /**
-     * Constructs an unconfigured, empty label.
+     * Constructs an unconfigured, empty label. Depending on the label type, transformations are applied
+     * once when typesetting (static) or every frame during rendering (dynamic).
      * @param text - The text that is displayed by this label.
+     * @param type - Either static or dynamic. If static is used, all transformations are baked and modifications to
+     * on any of the label's transformations are expected to occur less often.
```
Comment (anne-gropler) [inline source/text/label.ts]: doc: use *to* or *on*, i think? but my English is not the yellow from the egg.

> (...) all transformations are baked and modifications **to on** any of the (...)

File: source/text/label.ts [hunk 5]
```diff
@@ -97,11 +107,20 @@ export class Label {
...
+    protected vertices(): GlyphVertices {
         const vertices = new GlyphVertices(this.length + this.ellipsis.length);
         return vertices;
     }
 
+
+    /**
+     * Interface intended to compute/update the label's static and dynamic transformations as well as invoking the
+     * typesetter in order to create the glyph vertices. Returns undefined, if previous vertices can be reused since no
+     * was required. Returns an empty GlyphVertices storage if label is invalid or cannot be rendered ...
```
Comment (anne-gropler) [inline source/text/label.ts]: > (...) since **no was** required (...)
missing word, maybe *typesetting*?

Comment (anne-gropler) [reply source/text/label.ts]: and thanks for adding the abstract super method! :)

File: source/text/labelgeometry.ts [hunk 6]
```diff
@@ -14,7 +18,11 @@ import { Initializable } from '../initializable';
  * ```
  * const labelGeometry = new LabelGeometry(this._context);
  * labelGeometry = new LabelGeometry(this._context);
- * const aVertex = this._program.attribute('a_quadVertex', 0);
+ * this._program.attribute('a_vertex', labelGeometry.vertexLocation);
+ * this._program.attribute('a_vertex', labelGeometry.vertexLocation);
+ * this._program.attribute('a_vertex', labelGeometry.vertexLocation);
+ * this._program.attribute('a_vertex', labelGeometry.vertexLocation);
+ * this._program.attribute('a_vertex', labelGeometry.vertexLocation);
```
Comment (anne-gropler) [inline source/text/labelgeometry.ts]: five times the same line?

File: source/text/labelgeometry.ts [hunk 7]
```diff
@@ -79,21 +115,23 @@ export class LabelGeometry extends Geometry {
...
-        /* origin */
-        this._buffers[2].attribEnable(indices[2], 3, gl.FLOAT, false, 3 * 4, 0, true, false);
-        gl2facade.vertexAttribDivisor(indices[2], 1);
-        /* tangent */
-        this._buffers[3].attribEnable(indices[3], 3, gl.FLOAT, false, 3 * 4, 0, true, false);
-        gl2facade.vertexAttribDivisor(indices[3], 1);
-        /* up */
-        this._buffers[4].attribEnable(indices[4], 3, gl.FLOAT, false, 3 * 4, 0, true, false);
-        gl2facade.vertexAttribDivisor(indices[4], 1);
+        this._vertices.attribEnable(this._vertexLocation,
```
Comment (anne-gropler) [inline source/text/labelgeometry.ts]: Since the parameter `indices` is unused now, can you specify that in the method doc?

File: source/text/labelgeometry.ts [hunk 8]
```diff
@@ -79,21 +115,23 @@ export class LabelGeometry extends Geometry {
...
+        this._origins.attribEnable(this._originLocation,
+            3, gl.FLOAT, false, 12, 0, true, false);
+        gl2facade.vertexAttribDivisor(this._originLocation, 1);
+
+        this._tangents.attribEnable(this._tangentLocation,
+            3, gl.FLOAT, false, 12, 0, true, false);
+        gl2facade.vertexAttribDivisor(this._tangentLocation, 1);
+        this._bitangents.attribEnable(this._bitangentLocation,
+            3, gl.FLOAT, false, 12, 0, true, false);
+        gl2facade.vertexAttribDivisor(this._bitangentLocation, 1);
```
Comment (anne-gropler) [inline source/text/labelgeometry.ts]: Is there a meaning behind the grouping (missing empty lines)? Same for `unbindBuffers()`.

File: source/text/labelgeometry.ts (hunk 7 above)
Comment (anne-gropler) [reply source/text/labelgeometry.ts]: Same for `unbindBuffers()`

File: source/text/labelgeometry.ts [hunk 9]
```diff
@@ -102,65 +140,159 @@ export class LabelGeometry extends Geometry {
...
+
         const gl = this.context.gl;
         const valid = super.initialize(
-            [gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER],
-            [aQuadVertex, aTexCoord, aOrigin, aTangent, aUp]);
+            [gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER]);
 
         /**
          * These vertices are equal for all quads. There actual position will be changed using
          * origin, tangent and up(-vector).
```
Comment (anne-gropler) [inline source/text/labelgeometry.ts]: up-vector is now called bitangent (although I don't like it to be bitangent ^^ but we should be consistent in the docs...)

File: source/text/labelgeometry.ts [hunk 10]
```diff
@@ -102,65 +140,159 @@ export class LabelGeometry extends Geometry {
...
+        this._texCoordLocation = texCoordLocation;
+        this._originLocation = originLocation;
+        this._tangentLocation = tangentLocation;
+        this._bitangentLocation = bitangentLocation;
+
         const gl = this.context.gl;
         const valid = super.initialize(
-            [gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER],
-            [aQuadVertex, aTexCoord, aOrigin, aTangent, aUp]);
+            [gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER, gl.ARRAY_BUFFER]);
```
Comment (anne-gropler) [inline source/text/labelgeometry.ts]: I am surprised that the compiler does not complain about the missing non-optional `indices` param here :D

File: website/js/website.js (hunk 1 above)
Comment (anne-gropler) [reply website/js/website.js]: `setInterval()` in `labelrenderer.ts` triggers new frame quite frequently. So if you remove the `setInterval` and change the `multiFrameNumber` back, you got again a very pleasant accumulated frame :)

File: source/text/glyph.vert (hunk 2 above)
Comment (anne-gropler) [reply source/text/glyph.vert]: Reason was that those terms are related to the glyphs on the baseline, i.e., "tangent" is the vector tangential to the baseline, "up" is the vector orthogonal to the baseline. Since those terms are only used in typesetting space (which is 2D), we could change "up"/"bitangent" to "orthogonal", but this would merely be a mathematical description and would not contribute to the comprehension of the code.

"bitangent" would mean "tangential to the baseline in two (or more) points", which really does not fit our meaning here, as it is mathimatically incorrect.

File: source/text/labelrenderpass.ts [hunk 11]
```diff
@@ -260,34 +244,38 @@ export class LabelRenderPass extends Initializable {
...
-        this._geometry3D.bind();
-        this._geometry3D.draw();
+        this._geometry.bind();
 
-        gl.uniformMatrix4fv(this._uViewProjection, gl.GL_FALSE, mat4.create());
+        for (let i = 0; i < this._labels.length; ++i) {
+            const label = this._labels[i];
+            const range = this._ranges[i];
+            if (range[0] === range[1]) {
+                continue;
```
Comment (anne-gropler) [inline source/text/labelrenderpass.ts]: so, it is an empty label? (just for my comprehension)

File: source/text/glyph.vert (hunk 2 above)
Comment (anne-gropler) [reply source/text/glyph.vert]: For consistency, in some docs we talk about "direction" and "up" (user perspective) and and some others we use "tangent" and "bitangent" (mathematical perspective).

File: source/text/label.ts [hunk 12]
```diff
@@ -266,7 +285,11 @@ export class Label {
      * texture atlas).
```
Comment (anne-gropler) [inline source/text/label.ts]: wrong doc: lineWidth is *not* expected in typesetting space unit, but in label.fontSizeUnit

File: source/text/typesetter.ts [hunk 13]
```diff
@@ -344,7 +346,7 @@ export class Typesetter {
     private static transform(label: Label, vertices: GlyphVertices, lines: Array<Line>): void {
```
Comment (anne-gropler) [inline source/text/typesetter.ts]: Who let the docs out?

File: source/text/glyphvertices.ts [hunk 14]
```diff
@@ -1,128 +1,145 @@
...
+     * @param numberOfGlyphs - Target number of glyphs to reduce the vertices to.
+     */
+    shrink(numberOfGlyphs: number): void {
+        const length = Math.max(0, numberOfGlyphs);
+
+        /* Utility that slices the underlying buffer to the required length in bytes, and then creates a new
+        float 32 array view to that buffer. */
+        const shrink = (array: Float32Array, length: number, components: number) =>
+            array = new Float32Array(
+                array.buffer.slice(0, length * array.BYTES_PER_ELEMENT * components),
```
Comment (bwasty) [inline source/text/glyphvertices.ts]: This doesn't seem to match the doc comment - `slice` always copies data (-> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer/slice).
For avoiding copies, use the 3-argument constructor (`new Float32Array(buffer [, byteOffset [, length]]);`).

File: source/text/labelrenderpass.ts (hunk 11 above)
Comment (cgcostume) [reply source/text/labelrenderpass.ts]: yes
